### DISCUSSION TITLE
v0.70.0-beta2024120501

### DISCUSSION THREAD

**Discussion by yhakbar** [gruntwork-io/terragrunt#3626] [Pre-release Feedback]
## Deprecation Notice

This release deprecates `terragrunt.hcl` as the idiomatic root of Terragrunt configurations.

Naming the file placed at the root of Terragrunt configurations `terragrunt.hcl` is a pattern that does not provide any functional advantage over calling it anything else (e.g. `root.hcl`), but does serve to provide additional confusion and risk to Terragrunt users, as they can easily use Terragrunt in such a way that it ends up treating the `terragrunt.hcl` file at the root as a unit instead of shared configuration.

This pattern is very longstanding in typical Terragrunt usage, so changing it can be difficult. To get the ball rolling on this, the behavior will be deprecated, but still supported for the near future (and for a pretty long time thereafter).

### How this will be achieved

Terragrunt will now emit warnings whenever it detects that users are following a pattern indicating usage of a root `terragrunt.hcl` file.

The warning will look like this:

```
16:04:54.555 WARN   Using `terragrunt.hcl` as the root of Terragrunt configurations is an anti-pattern, and no longer recommended. In a future version of Terragrunt, this will result in an error. You are advised to use a differently named file like `root.hcl` instead. For more information, see https://terragrunt.gruntwork.io/docs/migrate/migrating-from-root-terragrunt-hcl
```

You can take a look at what those docs will look like [here](https://github.com/gruntwork-io/terragrunt/blob/feat/deprecate-terragrunt-hcl-as-root/docs/_docs/05_migration_guides/migrating-from-root-terragrunt-hcl.md).

Those docs include information on a new Strict Control named [root-terragrunt-hcl](https://github.com/gruntwork-io/terragrunt/blob/feat/deprecate-terragrunt-hcl-as-root/docs/_docs/04_reference/strict-mode.md#root-terragrunt-hcl) that will allow users to opt-in for a breaking change preventing usage of a root `terragrunt.hcl`.

Using this strict control will also update defaults for commands like `scaffold` and `catalog` so that they don't assume usage of a root `terragrunt.hcl` file.

### Feedback Requested

You are encouraged to share your feedback on this pre-released change. You can either share your feedback via the associated GitHub Discussion, or you can share your feedback in our Discord.

**Full Changelog**: https://github.com/gruntwork-io/terragrunt/compare/v0.69.7...v0.70.0-beta2024120501

<hr /><em>This discussion was created from the release <a href='https://github.com/gruntwork-io/terragrunt/releases/tag/v0.70.0-beta2024120501'>v0.70.0-beta2024120501</a>.</em>

**Comment by keiransteele-phocas** [comment]
Hi, could ```find_in_parent_folders()``` not also be changed to look for "root.hcl" as well to ease the migration?

**Reply by yhakbar** [reply]
Hey @keiransteele-phocas ,

We were discussing this internally. The issue that I have with this, personally, is that I don't think it actually helps anyone.

We are still going to support having a `terragrunt.hcl` file as the root of Terragrunt configurations for a long time. We are merely going to be warning users that they should explicitly indicate where they are trying to include configurations from.

I personally don't think seeing this in `terragrunt.hcl` files is a good experience for new users, but I would love feedback on that:

```hcl
include "root" {
  path = find_in_parent_folders()
}
```

I know that when I first saw that (and frequently when I first show it to other users), I was confused as to what `find_in_parent_folders` was doing. I think it's clearer when indicating the name of the file that's being found:

```hcl
include "root" {
  path = find_in_parent_folders("root.hcl")
}
```

At the end of the day, there's nothing special about the "root" Terragrunt configuration. It's just another configuration being included that could be called whatever a user wants.

**Reply by Jpk518** [reply]
Hey @yhakbar. Just want to give +1 to the original idea here. 

Having to update every single states `terragrunt.hcl` file for this change can be quite the undertaking in any brownfield codebase. Being able to transparently update the root file name from `terragrunt.hcl -> root.hcl` without needing to modify every leaf `terragrunt.hcl` would be more practical reduce migration pain.

Also this might not be the best place for discussion on the syntax, but I think while making the leaf node include statement more explicit has benefits in readability, I also think it adds more framework knowledge requirements to the leaf nodes and also increases the amount of repeated code. The consumers/leaf nodes probably shouldn't need to define such knowledge from the framework in general, and it still might be best to have this as concise and abstracted from these consumers as possible, with the ability to make it explicit if possible.

```
# Old
include {
  path = find_in_parent_folders()
}

# New
include "root" {
  path = find_in_parent_folders("root.hcl")
}
```

The parallel that comes to my mind although it might not be an entirely accurate analogy, is I wouldn't want to define the Go path or architecture in each file, but rather once in the respective env var/project configuration.

**Reply by yhakbar** [reply]
@Jpk518 
The fact of the matter is, the "root" include has felt like an implementation detail of Terragrunt for a long time for historical reasons, but it really isn't any more special than any other include.

Nothing about the file that you include from the root of your repository works different than other files you include. By having special syntax for `find_in_parent_folders()` to find the `root.hcl` file, we're claiming that there's something special about that file. There isn't.

It's totally valid, for example, for users to always have two includes, one for remote state and one for provider configurations in every unit. They might choose to have those files in `backend.hcl` and `provider.hcl` at the root of their repository and not have _any_ file named `root.hcl`. Users might also decide to have _multiple_ `root.hcl` files throughout their codebase, one in each directory that corresponds to a different cloud provider (e.g. `aws/root.hcl`, `azure/root.hcl`, `gcp/root.hcl`).

By making it clear that there is no special file that you _should_ put at the top of your repository with special tooling to discover it, we are making it clear that it's a _choice_.

To be clear, this won't be a breaking change for a _long_ time (we are guaranteeing that it will not be a breaking change before 2.0). You do not have to make changes in your brownfield codebases in the near term, and you will have plenty of time and tooling to adopt to new patterns before it becomes a requirement that you adopt to this change.

**Comment by VESHR** [comment]
Hi,the suggestion is to only name root file anything other than 'terragrunt.hcl' but child can still have 'terragrunt.hcl'?

**Reply by yhakbar** [reply]
Yup! The suggested name is `root.hcl` (that's what all the documentation has been updated to), but units (the leaf directories with `terragrunt.hcl` files in them that relate to particular infrastructure) should still be named `terragrunt.hcl`.

You can read more about it [here](https://terragrunt.gruntwork.io/docs/migrate/migrating-from-root-terragrunt-hcl/).
